src/forbidden_tree.jl

```diff
@@ -0,0 +1,188 @@
+        end
+        return forbidden_domain, vars
+    end
+end
+
+# Matching RuleNode with ConstraintMatchVar
+function _match_expr_with_hole(
+    ::RuleNode, 
+    ::ConstraintMatchVar, 
+    ::Vector{Int}
+)::Union{Tuple{Union{Int, Symbol}, Dict{Symbol, RuleNode}}, Nothing}
+    return nothing
+end
+
+# Matching Hole with ConstraintMatchNode
+function _match_expr_with_hole(
+    ::Hole, 
+    cmn::ConstraintMatchNode, 
+    hole_location::Vector{Int}
+ )::Union{Tuple{Union{Int, Symbol}, Dict{Symbol, RuleNode}}, Nothing}
+    if hole_location == [] && cmn.children == []
+        return cmn.rule_ind, Dict()
+    end
+    return nothing
+end
+
+# Matching Hole with ConstraintMatchVar
+function _match_expr_with_hole(
+    ::Hole, 
+    cmn::ConstraintMatchVar, 
+    hole_location::Vector{Int}
+ )::Union{Tuple{Union{Int, Symbol}, Dict{Symbol, RuleNode}}, Nothing}
+    if hole_location == []
+        return cmn.var_name, Dict()
+    end
+    return nothing
+end
+
+# Matching RuleNode with ConstraintMatchNode
+"""
+Tries to match RuleNode `rn` with ConstraintMatchNode `cmn`.
+Returns a dictionary of values assigned to variables if the match is successful.
+"""
+function _match_expr(rn::RuleNode, cmn::ConstraintMatchNode)::Union{Dict{Symbol, RuleNode}, Nothing}
+    if rn.ind ≠ cmn.rule_ind || length(rn.children) ≠ length(cmn.children)
+        return nothing
+    else
+        # Root nodes match, now we check the child nodes
+        vars = Dict()
+        for varsᵢ ∈ map(ab -> _match_expr(ab[1], ab[2]), zip(rn.children, cmn.children)) 
+            if varsᵢ ≡ nothing 
+                # unsuccessful match
+                return nothing
+            end
+            # Check if another argument already assigned the same variables
+            for (k, v) ∈ varsᵢ
+                if k ∈ keys(vars) && v ≠ vars[k]
+                    return nothing
+                end
+                vars[k] = v
+            end
+        end
+        return vars
+    end
+end
+
+# Matching RuleNode with ConstraintMatchVar
+_match_expr(rn::RuleNode, cmn::ConstraintMatchVar)::Union{Dict{Symbol, RuleNode}, Nothing} = Dict(cmn.var_name => rn)
+
+# Matching Hole
+_match_expr(::Hole, ::AbstractConstraintMatchNode)::Union{Dict{Symbol, RuleNode}, Nothing} = nothing
+
+"""
+Retrieves a rulenode at the original location by reference. 
+"""
+function get_node_at_location(root::RuleNode, location::Vector{Int})
+    if location == []
+        return root
+    else
+        return get_node_at_location(root.children[location[1]], location[2:end])
+    end
+end
+
+function get_node_at_location(root::Hole, location::Vector{Int})
+    if location == []
+        return root
+    end
+    return nothing
+end
+
+
+"""
+Propagates the ForbiddenTree constraint.
+It removes the elements from the domain that would complete the forbidden tree.
+"""
+function propagate(c::ForbiddenTree, ::Grammar, context::GrammarContext, domain::Vector{Int})
+    for i ∈ 0:length(context.nodeLocation)
```

Comment:
should we check everything along the path?
If it is a large expression, we would be checking lots of stuff again and again because `propagate_constraints` gets  called every time an expression is expanded.

Reply:
I think it will be quite difficult to reduce this. Most of our constraints will only have a depth of 2 or 3. However, if our constraint is e.g. removing `a+a` from the search space, and the root node of our expression is `+`, we still need to check this even though the depth of the node we are expanding might be much greater than 2, since this expansion could make the left and right subtree of the root node equal.
Afaik, this is the only reason why we need to match at every layer. 

An important observation is that this equality only works if the left subtree doesn't have any holes and the right subtree only has the hole we are expanding. If there are additional holes, we do not know how those will be expanded and as such we cannot reason about the equality of the trees. 
I don't think we gain much from checking the number of holes during the match; we still have to do this every iteration and it traverses the entire expression. 
We could however keep track of the number of holes that are (indirect) children for each rulenode. It would not be expensive to update this during expansion in the search.
Taking into consideration that we have to make some bigger changes in both Search and Grammar, I think we should turn this into a separate issue.

Reply:
good point!

Would it help if we re-envision `propagate_constraints` to also return an additional argument which are the new constraints to impose? That way, we can propagate the equality constraints 'lower' and only keep checking the children.

For instance, if we have `H + 2 x T` (where `H` and `T` are holes), we can split the constraint further down by stating that `H` has to be different from `2xT` and we never have to check the root node with `+` anymore; we can just return a new constraint. This would perhaps require us how to provide references to the nodes in a partial expression.

Reply:
Yes, this would definitely help. I think we should create a new class of local constraints. 
I can think of multiple ways of doing this, but I think the best way is to store those constraints together with the partial tree, and not in specific RuleNodes. For example, a `NotEquals` constraint would have two references to nodes that shouldn't equal each other. The benefit of this approach is that we do not have to explore the tree to find the local constraints, which can be beneficial in larger trees. Furthermore, we keep the RuleNodes as simple as possible.

There are two possibilities for referencing specific RuleNodes:
  - Currently, we reference to a RuleNode/Hole as a sequence of indices, that represents the child index at every parent. We could use this path to only propagate constraints for which one of the paths matches the beginning of the path to the hole we expanded.
  - For this application, we can also use actual references to the RuleNode. This could save us a lot of tree traversal.
I think we should try to use both options in parallel to get the benefits from both. 

I turned this into an issue: https://github.com/Herb-AI/HerbConstraints.jl/issues/7